### PR TITLE
[REVIEW] Provide External Memory Management Plugin for Numba

### PR DESCRIPTION
**Pull request by gmarkall**
The External Memory Management (EMM) Plugin interface is as-documented in https://github.com/gmarkall/nbep-7 - providing a plugin forces Numba to do all allocation through the plugin, so in this case all allocations use RMM instead of Numba directly using the driver API.

There is a corresponding PR in Numba: https://github.com/numba/numba/pull/5328

At the moment this illustrates what the changes would look like - all tests were passing before I went on leave, now after rebasing on branch-0.13 there are a couple of segfaults in the Numba test suite, I assume because of some differences in the `DeviceBuffer` compared to using the "raw" `rmm_alloc` - I'm looking into these.

This PR makes the presumption that all the functions that are reimplementations of Numba functions (`device_array`, `get_ipc_handle`), etc., would be removed since the Numba ones can be used directly when the EMM plugin is in use. Does this match the expectations of others? Would we want to keep around `rmm.device_array`, etc.?

Would we want to always force Numba to use RMM for allocations if RMM is imported (e.g. by calling `numba.cuda.cudadrv.set_memory_manager()` when `rmm` is imported), or is this a choice best left to the user of RMM?

All RMM tests pass with these changes.

I'll be updating the PR as I'm working on things - I think what's needed here in RMM is:

- Fix the segfaults in the Numba test suite (possibly / probably related to IPC)
- Documentation and testing
- Remove the implementation of `get_info` if it is removed from RMM (c.f. https://github.com/rapidsai/rmm/issues/305)

Any comments / questions welcomed at any time - many thanks in anticipation!

### CODE REVIEW

File: python/rmm/rmm.py [hunk 1]
```diff
@@ -233,6 +138,78 @@ def get_info(stream=0):
...
+            return IpcHandle(memory, ipchandle, memory.size, source_info,
+                             offset=offset)
+
+        def get_memory_info(self):
+            return get_info()
+
+        def initialize(self):
+            super().initialize()
+            if not self._initialized:
+                reinitialize(logging=self._logging)
```
Comment (kkraus14) [inline python/rmm/rmm.py]: How can we pass additional kwargs here? I.E. if we want to use a pool or managed memory?

File: python/rmm/rmm.py [hunk 2]
```diff
@@ -233,6 +138,78 @@ def get_info(stream=0):
...
+            # at the C++ layer for this, and also not rely on borrowing bits of
+            # Numba internals to initialise ipchandle.
+            ipchandle = (ctypes.c_byte * 64)()  # IPC handle is 64 bytes
+            cuda.cudadrv.memory.driver_funcs.cuIpcGetMemHandle(
+                ctypes.byref(ipchandle),
+                memory.owner.handle,
+            )
+            source_info = cuda.current_context().device.get_device_identity()
+            ptr = memory.device_ctypes_pointer.value
+            offset = librmm.rmm_getallocationoffset(ptr, stream)
```
Comment (jrhemstad) [inline python/rmm/rmm.py]: @harrism I forgot about `rmmGetAllocationOffset`. What do you think the future of this API is given the changes we're proposing in https://github.com/rapidsai/rmm/issues/301 https://github.com/rapidsai/rmm/issues/302 https://github.com/rapidsai/rmm/issues/305.

Looking at the implementation, it's independent of any resource since it uses the [CUDA driver API directly](https://github.com/rapidsai/rmm/blob/branch-0.13/src/rmm.cpp#L99). So we don't have to worry about burdening resources with implementing this functionality.

Comment (gmarkall) [reply python/rmm/rmm.py]: I'm using the CUDA driver here because it was a route I could see to get this to work - I found that some of the Numba tests didn't work with the existing implementation from the deleted `get_ipc_handle` above (I can't recall the exact details unfortunately). It feels as though it would be cleaner from this side if resources did implement this functionality - I don't feel confident this implementation will work robustly with all possible resource implementations, though I'm not familiar enough with them to state specifically what issues could arise.

Comment (jrhemstad) [reply python/rmm/rmm.py]: > I'm using the CUDA driver here because it was a route I could see to get this to work 

To be clear, I was referring to the use of the driver API in the implementation of `rmmGetAllocationOffset` which uses `cuMemGetAddressRange`.

This is necessary because [`cudaIpcGetMemHandle`](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g8a37f7dfafaca652391d0758b3667539) requires the _base_ pointer. When using a pool resource, a given pointer `p` resides somewhere within a larger `cudaMalloc'd` chunk of memory pointed to by `b`. In order to get an IPC handle for `p` we need to get `b` to pass into `cudaIpcGetMemHandle`.

Comment (harrism) [reply python/rmm/rmm.py]: @jrhemstad yes I had thought about this yesterday when I first saw this PR. I think when we  remove `rmmInitialize`, `rmmAlloc` and friends  that `rmmGetAllocationOffset` should change to the following free function.

`ptrdiff_t  rmm::allocation_offset(void* ptr);`

Note no need for a stream (currently unused), and throws exceptions rather than returns errors. Python code would use this rather than calling the CUDA driver API directly.

Comment (jrhemstad) [reply python/rmm/rmm.py]: Makes sense. Would RMM still be able to be header only in that case? Otherwise the driver lib will need to be linked in to the user app to support the `cuMemGetAddressRange` call.

Comment (harrism) [reply python/rmm/rmm.py]: Ugh. That's true.

Comment (harrism) [reply python/rmm/rmm.py]: Darn only driver API function we use! I guess that's not that bad... and we may choose to use some of the new virtual memory management APIs in the future, which are also driver only...

File: python/rmm/rmm.py (hunk 1 above)
Comment (gmarkall) [reply python/rmm/rmm.py]: I realized it was a mistake to be calling `reinitialize()` in the `reset()` function anyway - `reset` should only clear all allocations from the current context, not reset the state of the entire memory manager.

If you do want to call `reinitialize()`, I'd expect that it would be done outside of the Numba EMM plugin interface. For example, at program startup you might:

1. Call `reinitialize` with approprate parameters for a pool, or managed memory.
2. Set the Numba memory manager (e.g. calling `rmm.use_rmm_for_numba()`.
3. Start using Numba (allocating device arrays, calling kernels, etc.)

Reinitializing RMM in the middle of program execution (if you wanted different settings) should also work OK, as long as you have reset any contexts that you were using first (so that you don't have allocations from before re-initialization lying around).

I'll push a commit shortly which removes this call to `reinitialize` from `reset`.

**Comment by gmarkall** [comment]
Base of the PR updated to branch-0.14.

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by gmarkall** [comment]
Now that Numba 0.49 has been released with the External Memory Management plugin, I have updated this PR to match the released version. I think this is ready for some further review, but I need a little advice in some areas (perhaps from @kkraus):

## Local testing successes

 I have tested locally:

- The RMM testsuite (`test-rapids-python` in rapids-compose)
- The Numba testsuite with RMM as the allocator (with `NUMBA_CUDA_MEMORY_MANAGER=rmm python -m numba.runtests numba.cuda.tests`) 
- The cuDF testsuite with the code in https://github.com/rapidsai/cudf/pull/5042 (`test-cudf-python` in rapids-compose)

These tests were run with Numba modified so that it was incapable of allocating memory itself, to prove that all allocations were made through RMM - I commented out the reference to the `cuMemAlloc` driver function, so it would crash if it ever accidentally attempted to allocate memory itself.

## Local testing fails

I notice that other libraries (e.g. cuGraph) can fail because of making use of `device_array_from_ptr` which is removed - I understand there is some discussion about what this should be replaced with, but I'm not sure what the correct alternative is so I have not yet lined up PRs to fix those - is there a straightforward change I can make to replace these at present? Or is it likely that for this PR to be merged, `rmm.device_array_from_ptr` will need to be reinstated?

## CI testing

The environment for building on CI always picks Numba 0.48, so the tests are failing - is there a way to allow it to use 0.49?

## Testing with IPC

IPC is tested to the extent that the Numba testsuite tests IPC. I'm not sure that my local testing has exercised IPC very hard though - at one point the IPC implementation in this branch was broken, which I could tell from the Numba IPC tests failing, but all the cuDF tests still seemed to pass - is there some other test suite I should run to stress IPC a bit more (e.g. something involving dask?).

**Comment by gmarkall** [comment]
> The environment for building on CI always picks Numba 0.48, so the tests are failing - is there a way to allow it to use 0.49?

Just to clarify this: since I posted the comment I see it's only the `gpu-test` checks that have this problem - the `conda-build` tests seem to pick up Numba 0.49 without issue.

**Comment by gmarkall** [comment]
For now I've pushed a change that manually installs Numba 0.49 for testing - I will be able to remove this when https://github.com/rapidsai/gpuci-build-environment/pull/75 is merged. Other than that, I think this is ready for review @harrism @jrhemstad @kkraus14.

Since the previous comment https://github.com/rapidsai/rmm/pull/317#issuecomment-620790447 I've also tested with dask-cudf locally, with no issues. Some outstanding questions I have:

- Does `device_array_from_ptr` need reinstating in this PR so that users of it (e.g. cuGraph) can continue to use it? Or is a fix to replace it needed in cuGraph?
- Are the dask-cudf tests sufficient to assume confidence in the implementation of IPC handles?

**Comment by kkraus14** [comment]
> * Are the dask-cudf tests sufficient to assume confidence in the implementation of IPC handles?

No, the dask-cudf tests typically either just use the threaded scheduler or use TCP for communication.


> * Does `device_array_from_ptr` need reinstating in this PR so that users of it (e.g. cuGraph) can continue to use it? Or is a fix to replace it needed in cuGraph?

They should use `rmm._DevicePointer` instead most likely but I believe they're in progress of removing any raw allocations in general. @BradReesWork can loop in the right person(s) to answer this.

**Comment by gmarkall** [comment]
> > * Are the dask-cudf tests sufficient to assume confidence in the implementation of IPC handles?
> 
> No, the dask-cudf tests typically either just use the threaded scheduler or use TCP for communication.

OK - is there anything else I can/should do to exercise the IPC code? Although Numba has some IPC tests, they're not particularly exhaustive. They do all pass when using the RMM plugin, but I can imagine they won't hit many edge cases.

**Comment by kkraus14** [comment]
> OK - is there anything else I can/should do to exercise the IPC code? Although Numba has some IPC tests, they're not particularly exhaustive. They do all pass when using the RMM plugin, but I can imagine they won't hit many edge cases.

I believe cuGraph has code that heavily relies on IPC so it may get tested via that? Otherwise UCX-py under the hood relies on CUDA IPC (at the C/C++ level, not at the Python level) where it will get tested via that as well.

**Comment by awthomp** [comment]
@kkraus14 
> cc @dantegd for cuML, @Iroy30 / @afender for cuGraph, @thomcom for cuSpatial, @awthomp for cuSignal

I don't see any blockers in this PR for cuSignal.

File: ci/gpu/build.sh [hunk 3]
```diff
@@ -63,6 +63,7 @@ if hasArg --skip-tests; then
 else
     logger "Installing extra test dependencies..."
     conda install "cupy>=6.0.0"
+    conda install "numba>=0.49.0"
```
Comment (dillon-cullinan) [inline ci/gpu/build.sh:66]: This should be on the same line as the other conda install. It will save time solving the environment.

Comment (kkraus14) [reply ci/gpu/build.sh:66]: @dillon-cullinan this is just temporary as we bump the gpuci-build-environment Dockerfile anyway 😄. We can likely remove both installs after that!

**Comment by Iroy30** [comment]
> > OK - is there anything else I can/should do to exercise the IPC code? Although Numba has some IPC tests, they're not particularly exhaustive. They do all pass when using the RMM plugin, but I can imagine they won't hit many edge cases.
> 
> I believe cuGraph has code that heavily relies on IPC so it may get tested via that? Otherwise UCX-py under the hood relies on CUDA IPC (at the C/C++ level, not at the Python level) where it will get tested via that as well.

Currently in cugraph is steering away from the fraction of code that uses IPC

File: python/rmm/rmm.py [hunk 4]
```diff
@@ -130,107 +130,72 @@ def csv_log():
...
+            ctypes.byref(ipchandle), memory.owner.handle,
+        )
+        source_info = cuda.current_context().device.get_device_identity()
+        ptr = memory.device_ctypes_pointer.value
+        offset = librmm.rmm_getallocationoffset(ptr, 0)
+        return IpcHandle(
+            memory, ipchandle, memory.size, source_info, offset=offset
+        )
 
+    def get_memory_info(self):
```
Comment (shwina) [inline python/rmm/rmm.py:187]: Just curious if this is essential functionality or a nice-to-have?

I'm working on removing the use of deprecated RMM APIs in https://github.com/rapidsai/rmm/pull/361, `rmmGetInfo()` being one of those.

Further, the new `memory_resource` classes don't all support querying memory information, so depending on the memory resource being used, we may or may not be able to offer this functionality.

cc: @jrhemstad

Comment (gmarkall) [reply python/rmm/rmm.py:187]: Just a nice-to-have - it's allowed to raise `NotImplementedError`: http://numba.pydata.org/numba-doc/latest/cuda/external-memory.html#numba.cuda.BaseCUDAMemoryManager.get_memory_info

Comment (shwina) [reply python/rmm/rmm.py:213]: Thanks!

**Comment by gmarkall** [comment]
Because `device_array_from_ptr` doesn't have RMM do any allocation there's nothing mutually exclusive about having it coexist with using RMM for Numba. I've reinstated it in this PR as the cuGraph issue https://github.com/rapidsai/cugraph/issues/841 is in the 0.15 milestone - that way, we can remove it later when the cuGraph issue is resolved, rather than needing that change for this PR to go in.

**Comment by gmarkall** [comment]
@kkraus14 Good point - I've reinstated all the functions that I originally removed:

- `_make_finalizer`
- `device_array`
- `device_array_like`
- `get_ipc_handle`
- `to_device`

To test these and the use of `numba.cuda` for allocations, I added some extra parameterizations to the tests, which can be removed when these functions are eventually removed.

This tests OK with cuDF branch-0.14 and also with https://github.com/rapidsai/cudf/pull/5042 - I believe this PR should now be able to be merged without requiring any changes in any library using RMM.

There are still a couple of blockers for merging, AIUI:

- Merge of https://github.com/rapidsai/gpuci-build-environment/pull/75 (approved but not yet merged)
- Merge of https://github.com/rapidsai/integration/pull/31 (approved but not yet merged)
- Removal of https://github.com/rapidsai/rmm/pull/317/files#diff-e19d03ed8e05824bbc937b07c4fddd6fR66 once the other two PRs are merged.
